### PR TITLE
#1084: decode body if base64 is enable

### PR DESCRIPTION
**Pull request by npeters**
*Issue #1084:*

*Description of changes:*
if isBase64Encoded == true then AwsStringHttpProcessingUtils must decode the body

By submitting this pull request

- [X] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [X] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

**Comment by npeters** [comment]
by default the content-type must be application/json

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java [hunk 1]
```diff
@@ -180,4 +197,36 @@ private static <T> T readValue(String json, Class<T> clazz, ObjectMapper mapper)
 		}
 	}
 
+    static final String HEADER_KEY_VALUE_SEPARATOR = "=";
+    static final String HEADER_VALUE_SEPARATOR = ";";
+    static final String ENCODING_VALUE_KEY = "charset";
+    static protected Charset parseCharacterEncoding(String contentTypeHeader) {
```
Comment (deki) [inline aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:203]: Similar code is in place here https://github.com/aws/serverless-java-container/blob/f85319ec59ad2b452bdef8db012d19d74510bbc3/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java#L372 
wouldn't it make more sense to centralize?

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java [hunk 2]
```diff
@@ -120,10 +122,17 @@ private static HttpServletRequest generateRequest1(String request, Context lambd
...
-			httpRequest.setContent(v1Request.getBody().getBytes(StandardCharsets.UTF_8));
-		}
+        if (StringUtils.hasText(v1Request.getBody())) {
+			if (v1Request.getHeaders().get(HttpHeaders.CONTENT_TYPE)==null) {
+				httpRequest.setContentType("application/json");
+			}
+            if (v1Request.isBase64Encoded()) {
+                httpRequest.setContent(Base64.getMimeDecoder().decode(v1Request.getBody()));
+            } else {
+                Charset charseEncoding = parseCharacterEncoding(v1Request.getHeaders().get(HttpHeaders.CONTENT_TYPE));
```
Comment (deki) [inline aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:132]: now it's parsed, wouldn't it make sense to pass using `setContentType` as well?

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java [hunk 3]
```diff
@@ -149,11 +158,19 @@ private static HttpServletRequest generateRequest2(String request, Context lambd
...
 		
 		v2Request.getHeaders().forEach(httpRequest::setHeader);
-		
-		if (StringUtils.hasText(v2Request.getBody())) {
-			httpRequest.setContentType("application/json");
-			httpRequest.setContent(v2Request.getBody().getBytes(StandardCharsets.UTF_8));
-		}
+
+
+        if (StringUtils.hasText(v2Request.getBody())) {
```
Comment (deki) [inline aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:163]: replace duplicate code with extracted method from above

**Review by deki — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. Agree that this needs to be fixed. Could you reduce the duplicated code?
We plan to have a release early next week and it would be great to include it.

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java (hunk 1 above)
Comment (npeters) [reply aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:203]: The question for me is: where to centralize? ;) 

maybe a new class com.amazonaws.serverless.proxy.internal.HttpUtils? like SecurityUtils.
I tried looking at the code, but couldn't find anything to consider other than this method 

but it’s your project that you decide

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java [hunk 4]
```diff
@@ -120,10 +122,17 @@ private static HttpServletRequest generateRequest1(String request, Context lambd
 			MultiValueMapAdapter headers = new MultiValueMapAdapter(v1Request.getMultiValueHeaders());
 			httpRequest.setHeaders(headers);
 		}
-		if (StringUtils.hasText(v1Request.getBody())) {
-			httpRequest.setContentType("application/json");
-			httpRequest.setContent(v1Request.getBody().getBytes(StandardCharsets.UTF_8));
-		}
+        if (StringUtils.hasText(v1Request.getBody())) {
```
[the remark this answers is not in the document]
Comment (npeters) [reply aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:125]: I did it

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java (hunk 1 above)
Comment (deki) [reply aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:222]: Yes it has to be a new class as we have two separate request flows (old request flow without Spring Cloud functions and new flow with Spring Cloud functions). I'm fine with the utils approach...

File: aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java (hunk 3 above)
Comment (deki) [reply aws-serverless-java-container-springboot3/src/main/java/com/amazonaws/serverless/proxy/spring/AwsSpringHttpProcessingUtils.java:165]: @npeters the following lines here can now be removed, can't they?

**Comment by deki** [comment]
Thank you very much for your contribution, @npeters! The changes have been merged and will be part of the next release.

**Pull request merged**